File: wowfoot-import/ruby/json.rb
# helpers
class String
	def beginsWith(other)
		os = other.to_s
		return self[0,os.length] == (os)
	end
end

def matchchar(c, m)
	s = ' '
	s[0] = c
	return s.match(m) != nil
end

def isdigit(c)
	return c == '-'[0] || matchchar(c, /\d/)
end

def isspace(c)
	matchchar(c, /\s/)
end

def isletter(c)
	matchchar(c, /[A-Za-z]/)
end

class JsonParser
	def self.parse(line, pos)
		@@line = line
		@@pos = pos
		begin
			return parseValue
		rescue
			p "line: " + @@line
			puts "length: #{@@line.length}"
			puts "pos #{@@pos}"
			puts "place: #{@@line[@@pos, @@line.length]}"
			raise
		end
	end
	
	def self.parseValue
		#puts 'pV'
		skipWhitespace
		c = @@line[@@pos,1]
		return parseArray if(c == '[')
		return parseObject if(c == '{')
		return parseString(c) if(c == '"' || c == "'")
		return parseInt if(isdigit(c[0]))
		error
	end
	
	def self.parseString(delimiter)
		token(delimiter)
		s = ''
		while(@@line[@@pos,1] != delimiter)
			if(@@line[@@pos,1] == "\\")
				@@pos += 1
			end
			s += @@line[@@pos,1]
			@@pos += 1
		end
		@@pos += 1
		return s
	end
	
	def self.parseInt
		pos = @@pos
		while(isdigit(@@line[@@pos]))
			@@pos += 1
		end
		len = @@pos - pos
		return @@line[pos, len].to_i
	end
	
	# returns an array of values (integers, strings, hashes, arrays).
	def self.parseArray
		token('[')
		#puts 'pA'
		arr = []
		while(true)
			arr << parseValue
			
			if(next?(']'))
				next
			else
				break
			end
		end
		return arr
	end
	
	# returns a hash (string => value). that's what a JSON object is.
	def self.parseObject
		token('{')
		hash = {}
		while(true)
			skipWhitespace
			namePos = @@pos
			while(isletter(@@line[@@pos]))
				@@pos += 1
			end
			nameLen = @@pos - namePos
			name = @@line[namePos, nameLen]
			token(':')
			hash[name.to_sym] = parseValue
			
			if(next?('}'))
				next
			else
				break
			end
		end
		return hash
	end

	# helpers
	def self.error
		p @@line[@@pos-1, @@line.length]
		raise hell
	end
	
	def self.next?(breakToken)
		c = @@line[@@pos,1]
		@@pos += 1
		return true if(c == ',')
		return false if(c == breakToken)
		error
	end
		
	def self.skipWhitespace
		while(isspace(@@line[@@pos]))
			@@pos += 1
		end
	end
	def self.token(t)
		skipWhitespace
		error unless(@@line[@@pos, t.length] == t)
		@@pos += t.length
	end
end

File: wowfoot-import/src/json.cpp
#include "json.h"
#include <vector>
#include <assert.h>

using namespace std;

// Classes

class JsonValueImpl : public JsonValue {
public:
	string mValue;
public:
	const string& toString() const { return mValue; }
};

class JsonArrayImpl : public JsonArray {
public:
	vector<JsonValue*> mVec;
public:
	virtual ~JsonArrayImpl() {
		for(size_t i=0; i<mVec.size(); i++) {
			delete mVec[i];
		}
	}
	const JsonArray* toArray() const { return this; }
	virtual size_t size() const { return mVec.size(); }
	virtual const JsonValue& operator[](size_t i) const { return *mVec[i]; }
};

class JsonObjectImpl : public JsonObject {
public:
	Map mMap;
public:
	virtual ~JsonObjectImpl() {
		for(Map::iterator itr=mMap.begin(); itr != mMap.end(); ++itr) {
			delete itr->second;
		}
	}
	const JsonObject* toObject() const { return this; }
	virtual const JsonValue& operator[](const string& key) const {
		Map::const_iterator itr = mMap.find(key);
		if(itr == mMap.end())
			abort();
		else
			return *itr->second;
	}
	
	virtual iterator begin() const { return mMap.begin(); }
	virtual iterator end() const { return mMap.end(); }
};

// Statics
const char* sText;

// Functions
static JsonValue* parseValue();
static JsonArray* parseArray();
static JsonObject* parseObject();
static JsonValue* parseInt();
static JsonValue* parseString(char delim);
static void error() __attribute((noreturn));
static void skipWhitespace();
static void token(char);
static bool isJsonDigit(char);
static bool nextInList(char breaker);

JsonValue* parseJson(const char* text) {
	sText = text;
	return parseValue();
}

static JsonValue* parseValue() {
	skipWhitespace();
	char c = *sText;
	if(c == '[')
		return parseArray();
	if(c == '{')
		return parseObject();
	if(c == '\'' || c == '"')
		return parseString(c);
	if(isJsonDigit(c))
		return parseInt();
	error();
}

static JsonValue* parseString(char delim) {
	token(delim);
	string s;
	while(*sText != delim) {
		if(*sText == '\\')
			sText++;
		s += *sText;
		sText++;
	}
	sText++;
	JsonValueImpl* v = new JsonValueImpl;
	v->mValue = s;
	return v;
}

static JsonValue* parseInt() {
	string s;
	while(isJsonDigit(*sText)) {
		s += *sText;
		sText++;
	}
	JsonValueImpl* v = new JsonValueImpl;
	v->mValue = s;
	return v;
}
static int level = 0;
static JsonArray* parseArray() {
	level++;
	assert(level < 10);
	token('[');
	JsonArrayImpl* arr = new JsonArrayImpl;
	while(true) {
		if(*sText == ']')
			break;
		arr->mVec.push_back(parseValue());
		if(nextInList(']'))
			continue;
		else
			break;
	}
	level--;
	return arr;
}

static JsonObject* parseObject() {
	level++;
	assert(level < 10);
	token('{');
	JsonObjectImpl* obj = new JsonObjectImpl;
	while(true) {
		if(*sText == '}')
			break;
		skipWhitespace();
		string name;
		while(isalpha(*sText)) {
			name += *sText;
			sText++;
		}
		token(':');
		obj->mMap[name] = parseValue();
		
		if(nextInList('}'))
			continue;
		else
			break;
	}
	level--;
	return obj;
}

static bool isJsonDigit(char c) {
	return isdigit(c) || c == '-';
}

static void error() {
	printf("Json parse error. current text: (%i) %s\n", *sText, sText);
	abort();
}

static void skipWhitespace() {
	while(isspace(*sText)) {
		sText++;
	}
}

static void token(char c) {
	skipWhitespace();
	if(*sText != c)
		error();
	sText++;
}

static bool nextInList(char breakToken) {
	char c = *sText;
	sText++;
	if(c == ',')
		return true;
	if(c == breakToken)
		return false;
	error();
}
